Support killing of inboxed service invocations

tillrohrmann opened the pull request:
This PR adds the functionality to kill an inboxed service invocation. The way it works is the following:

1. Check whether an invocation is currently invoked.
2. If not, then check the inbox for the corresponding service invocation
3. If the inbox contains the service invocation, then fail it with a `KILLED_INVOCATION_ERROR` and delete the service invocation from the inbox.

This PR is based on #949.

This fixes #951.

slinkydeveloper commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -569,27 +608,38 @@ where
+        let length = invocation_metadata.journal_metadata.length;
+
+        self.try_send_failure_response(effects, &full_invocation_id, invocation_metadata, error);
+        self.complete_invocation(full_invocation_id, state, length, effects)
```
This method name is very misleading. Something like `end_invocation_lifecycle`, or somewhat like that, might make more obvious what it does.

tillrohrmann replied:
Will rename it.

slinkydeveloper reviewed (changes requested):
Can you add a test in state_machine/mod.rs?

tillrohrmann commented on crates/worker/src/partition/state_machine/command_interpreter.rs:
```diff
@@ -370,21 +377,53 @@ where
+
+        let invocation_metadata = InvocationMetadata::new(
+            fid.invocation_uuid,
+            JournalMetadata::new(0, span_context),
```
True. Let me see whether the underlying methods can be easily refactored.

tillrohrmann commented:
Thanks for the review @slinkydeveloper. I've updated the PR to resolve your comments and I added two tests for the `StateMachine` as well as for the `CommandInterpreter`. PTAL.

slinkydeveloper commented on crates/types/src/invocation.rs:
```diff
@@ -65,6 +65,21 @@ pub enum MaybeFullInvocationId {
     Full(FullInvocationId),
 }
 
+impl MaybeFullInvocationId {
+    pub fn into_invocation_id(self) -> InvocationId {
+        match self {
+            MaybeFullInvocationId::Partial(iid) => iid,
+            MaybeFullInvocationId::Full(fid) => InvocationId::from(fid),
+        }
+    }
+}
```
No need to add this method to duplicate the functionality of From/Into

tillrohrmann replied:
Will remove it.

tillrohrmann commented on crates/types/src/invocation.rs:
```diff
@@ -369,3 +369,20 @@ impl SpanRelation {
+
+#[cfg(any(test, feature = "mocks"))]
+mod mocks {
+    use crate::invocation::ServiceInvocation;
```
Will do.

tillrohrmann commented:
Thanks for the review @slinkydeveloper. I've rebased the PR and force pushed it. Once GH gives green light, I'll merge it.
